File: apps/macos/Mollotov/Handlers/BookmarkHandler.swift
import Foundation

/// API handler for bookmark CRUD: list, add, remove, clear.
struct BookmarkHandler {
    let context: HandlerContext

    func register(on router: Router) {
        router.register("bookmarks-list") { _ in await list() }
        router.register("bookmarks-add") { body in await add(body) }
        router.register("bookmarks-remove") { body in await remove(body) }
        router.register("bookmarks-clear") { _ in await clear() }
    }

    @MainActor
    private func list() async -> [String: Any] {
        successResponse(["bookmarks": BookmarkStore.shared.toJSON()])
    }

    @MainActor
    private func add(_ body: [String: Any]) async -> [String: Any] {
        guard let url = body["url"] as? String else {
            return errorResponse(code: "MISSING_PARAM", message: "url is required")
        }
        let title = body["title"] as? String ?? url
        BookmarkStore.shared.add(title: title, url: url)
        return successResponse(["bookmarks": BookmarkStore.shared.toJSON()])
    }

    @MainActor
    private func remove(_ body: [String: Any]) async -> [String: Any] {
        guard let idStr = body["id"] as? String, let id = UUID(uuidString: idStr) else {
            return errorResponse(code: "MISSING_PARAM", message: "id is required")
        }
        BookmarkStore.shared.remove(id: id)
        return successResponse(["bookmarks": BookmarkStore.shared.toJSON()])
    }

    @MainActor
    private func clear() async -> [String: Any] {
        BookmarkStore.shared.removeAll()
        return successResponse(["cleared": true])
    }
}

File: apps/android/app/src/main/java/com/kelpie/browser/handlers/BookmarkHandler.kt
package com.kelpie.browser.handlers

import com.kelpie.browser.browser.BookmarkStore
import com.kelpie.browser.network.Router
import com.kelpie.browser.network.errorResponse
import com.kelpie.browser.network.successResponse

class BookmarkHandler(
    private val ctx: HandlerContext,
) {
    fun register(router: Router) {
        router.register("bookmarks-list") { list() }
        router.register("bookmarks-add") { add(it) }
        router.register("bookmarks-remove") { remove(it) }
        router.register("bookmarks-clear") { clear() }
    }

    private suspend fun list(): Map<String, Any?> = successResponse(mapOf("bookmarks" to BookmarkStore.toJSON()))

    private suspend fun add(body: Map<String, Any?>): Map<String, Any?> {
        val url = body["url"] as? String ?: return errorResponse("MISSING_PARAM", "url is required")
        val title = body["title"] as? String ?: url
        BookmarkStore.add(title, url)
        return successResponse(mapOf("bookmarks" to BookmarkStore.toJSON()))
    }

    private suspend fun remove(body: Map<String, Any?>): Map<String, Any?> {
        val id = body["id"] as? String ?: return errorResponse("MISSING_PARAM", "id is required")
        BookmarkStore.remove(id)
        return successResponse(mapOf("bookmarks" to BookmarkStore.toJSON()))
    }

    private suspend fun clear(): Map<String, Any?> {
        BookmarkStore.clear()
        return successResponse(mapOf("cleared" to true))
    }
}
